libraries/persistent_store/src/model.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019-2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::format::Format;
+use crate::{StoreError, StoreRatio, StoreResult, StoreUpdate};
+use std::collections::{HashMap, HashSet};
+
+/// Models the mutable operations of a store.
+///
+/// The model doesn't model the storage and read-only operations. This is done by the driver.
+#[derive(Clone, Debug)]
+pub struct StoreModel {
+    /// Represents the content of the store.
+    map: HashMap<usize, Box<[u8]>>,
+
+    /// The modeled storage configuration.
+    format: Format,
+}
+
+/// Mutable operations on a store.
+#[derive(Clone, Debug)]
+pub enum StoreOperation {
+    /// Applies a transaction.
+    Transaction {
+        /// The list of updates to be applied.
+        updates: Vec<StoreUpdate>,
+    },
+
+    /// Deletes all keys above a threshold.
+    Clear {
+        /// The minimum key to be deleted.
+        min_key: usize,
+    },
+
+    /// Compacts the store until a given capacity is immediately available.
+    Prepare {
+        /// How much capacity should be immediately available after compaction.
+        length: usize,
+    },
+}
+
+impl StoreModel {
+    /// Creates an empty model for a given storage configuration.
+    pub fn new(format: Format) -> StoreModel {
+        let map = HashMap::new();
+        StoreModel { map, format }
+    }
+
+    /// Returns the modeled content.
+    pub fn map(&self) -> &HashMap<usize, Box<[u8]>> {
```

Comment:
This name might be confusing, since it conflicts with the commonly used `map` for i.e. `Result`. You have lines like:

> self.map.values().map(|x| self.entry_size(x)).sum();

And it's not super hard to understand, but it wouldn't hurt to rename the variable and getter?

Reply:
Right, I didn't see possible confusion here. I renamed to `content` which also fits the current documentation.